[ML-28124] Create script for generating model monitoring setup notebook (#48)
databricks/mlops-stack: {{cookiecutter.project_name}}/monitoring/README.md

> # Setup Model Monitoring

Reviewer: Should we consistently refer to it as `Data Monitoring` everywhere? That is the current name of the product, so it might be misleading to see model some places and data in others.

Author: Data monitoring is for general tables. This notebook is specifically for monitoring the inference log table.
There could be feature monitoring or other table monitoring in the future. Are you suggesting maybe unify them?

Author: data monitoring for inference table

> This directory contains setup scripts intended to generate notebook that can be uploaded to databricks workspace and execute

Reviewer 2: ```suggestion
This directory contains setup scripts intended to generate a notebook that can be uploaded to a Databricks workspace and execute
```

to set up model monitoring.

## Prerequisites

### Unity Catalog

> The inference table to be monitored must exist in Unity Catalog.

Reviewer 2: Nit: skip "inference"? Or use "table with model predictions" to be more clear?

Reviewer: +1

> If Unity Catalog is enabled for your workspace but the table exists in metastore, please run notebook code 

Reviewer 2: What does "run notebook code" mean?

> or considering create workflow to copy data to Unit Catalog.

Reviewer 2: ```suggestion
or consider creating a workflow to copy data to Unity Catalog.
```

> For example, the following code 

Reviewer 2: Expand this blurb? E.g., "the following code shows how to import your data to a Unity Catalog managed table"

 - load data from metastore example_project_predictions 
 - write data to Unity Catalog my_unity_catalog.my_schema.my_inference_table
 - convert prediction column to double type
 - convert inference_time_stamp to timestamp type
 - convert tpep_pickup_datetime to timestamp type
 - convert tpep_dropoff_datetime to timestamp type
```
df = spark.table("hive_metastore.default.example_project_predictions") \
        .withColumn("prediction",col("prediction").cast("double")) \
        .withColumn("inference_time_stamp",to_timestamp("inference_time_stamp")) \
        .withColumn("tpep_pickup_datetime",to_timestamp("tpep_pickup_datetime")) \
        .withColumn("tpep_dropoff_datetime",to_timestamp("tpep_dropoff_datetime"))
df.write.saveAsTable(
    name = "my_unity_catalog.my_schema.my_inference_table"
)
```

### Required columns

The inference table is required to contain the following columns
 - a column that contains the prediction results. For regression problems the type must be number type.

>  - a column with timestamp type that contains the time that the inference happens

Reviewer 2: ```suggestion
 - a column with timestamp type that contains the time that the inference happened.
```

>  - a column that contains the prediction results. For regression problems the type must be number type.
>  - a column with timestamp type that contains the time that the inference happens

Reviewer: nit wording suggestions
```suggestion
 - A column that contains the predictions. For regression problems, the column must be a subclass of `NumericType`.
 - A column with that contains the time of inference. This column must be of type `TimestampType`.
```

>  - a column with string type that represents the model version used to predict

Reviewer 2: ```suggestion
 - a column with string type that represents the model version used for inference.
```

>  - a column with string type that represents the model version used to predict

Reviewer: Note that this column does not need to be a string
```suggestion
 - A column that represents the model version used for inference.
```

## Usage

### 1. Run the scripts
From the monitoring directory, run:

```
python bootstrap.py
```

> A notebook `output/setup_model_monitoring.py` will be generated in the monitoring directory.

Reviewer 2: Why is this in an output subdirectory?

### 2. Update optional fields as necessary

The notebook `output/setup_model_monitoring.py` has required fields already filled in. Read the instructions 
in the notebook to update optional fields or add custom metrics.

### 3. Set up monitoring

> Upload the generated notebook file `output/setup_model_monitoring.py` to workspace and run it to generate 

Reviewer 2: Is this a one time thing? Curious if it's worth adding scaffolding in Stacks for it (e.g., terraform). Probably not worth it right?

> Upload the generated notebook file `output/setup_model_monitoring.py` to workspace and run it to generate 
> set up model monitoring. The user running the setup notebook must be owner of the inference table.

Reviewer: ```suggestion
Upload the generated notebook file `output/setup_model_monitoring.py` to workspace and run it to create and set up 
a monitor. The user running the setup notebook must be owner of the inference table.
```
